srml/support/procedural/tools/derive/src/lib.rs

```diff
@@ -0,0 +1,227 @@
+/// design but fast parse impl).
+#[proc_macro_derive(ParseStruct)]
+pub fn derive_parse_struct(input: TokenStream) -> TokenStream {
+	let syn::ItemStruct {
+		ident,
+		generics,
+		fields,
+		..
+	} = parse_macro_input!(input as syn::ItemStruct);
+	let field_names = {
+		let name = fields_idents(fields.iter().map(Clone::clone));
+		quote!{
+			#(
+				#name,
+				)*
+		}
+	};
+	let field = fields_idents(fields.iter().map(Clone::clone));
+	let tokens = quote! {
+		impl #generics syn::parse::Parse for #ident #generics {
+			fn parse(input: syn::parse::ParseStream) -> syn::parse::Result<Self> {
+				#(
+					let #field = input.parse()?;
+				)*
+				Ok(Self {
+					#field_names
+				})
+			}
+		}
+	};
+	tokens.into()
+}
+
+#[proc_macro_derive(ToTokensStruct)]
+pub fn derive_totokens_struct(input: TokenStream) -> TokenStream {
+	let syn::ItemStruct {
+		ident,
+		generics,
+		fields,
+		..
+	} = parse_macro_input!(input as syn::ItemStruct);
+	let fields = fields_access(fields.iter().map(Clone::clone));
+	let tokens = quote! {
+
+		impl #generics quote::ToTokens for #ident #generics {
+			fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+				#(
+					self.#fields.to_tokens(tokens);
+				)*
+			}
+		}
+
+	};
+	tokens.into()
+}
+
+
+/// self defined parsing enum, variant are tested in order of definition.
+/// Empty variant allways true.
+/// Please use carefully, this will fully parse successfull variant twice.
+#[proc_macro_derive(ParseEnum)]
+pub fn derive_parse_enum(input: TokenStream) -> TokenStream {
+
+	let syn::ItemEnum {
+		ident,
+		generics,
+		variants,
+		..
+	} = parse_macro_input!(input as syn::ItemEnum);
+	let variants = variants.iter().map(|v| {
+		let variant_ident = v.ident.clone();
+		let fields_build = if v.fields.iter().count() > 0 {
+			let fields_id = fields_idents(v.fields.iter().map(Clone::clone));
+			quote!( (#(#fields_id), *) )
+		} else {
+			quote!()
+		};
+
+		let fields_procs = fields_idents(v.fields.iter().map(Clone::clone))
+			.map(|fident| {
+				quote!{
+					let mut #fident = match fork.parse() {
+						Ok(r) => r,
+						Err(_e) => break,
+					};
+				}
+			});
+		let fields_procs_again = fields_idents(v.fields.iter().map(Clone::clone))
+			.map(|fident| {
+				quote!{
+					#fident = input.parse().expect("was parsed just before");
+				}
+			});
+
+		// double parse to update input cursor position
+		// TODO check next syn crate for a way
+		// to copy position/state from a fork
+		quote!{
+			let mut fork = input.fork();
+			loop {
+				#(#fields_procs)*
+				#(#fields_procs_again)*
+				return Ok(#ident::#variant_ident#fields_build);
+			}
+		}
+	});
+
+	let tokens = quote! {
+		impl #generics syn::parse::Parse for #ident #generics {
+			fn parse(input: syn::parse::ParseStream) -> syn::parse::Result<Self> {
+				#(
+					#variants
+				)*
+				// no early return from any variants
+				Err(
+					syn::parse::Error::new(
+						proc_macro2::Span::call_site(),
```

Comment:
`input.span()` would probably be better.

Reply:
I did not find a more convenient way to access the span of the ParseBuffer (does not implement 'Spanned') so I did use this (ParseBuffer initial scopte is initialized at this value).

Reply:
Yeah, you are right. It was a mistake to mark this in the first place. Sorry!